Private Constructor with parameter

I want to test a class that it has only exactly one constructor which is private and takes one argument. 

I thought

``` java
PrivateConstructorChecker.forClass(MyClass.class).check();
```

would do it but it gave me an exception back:

> java.lang.AssertionError: class onactivityresult.compiler.Parameter has non-default constructor with some parameters

So I guess this was a design decision. Would you be open for a new API to test the described behavior?

Maybe something alone the lines of

``` java
PrivateConstructorChecker.forClass(MyClass.class).expectWithParameters(...).check();
```

Obviously `expectWithParameters` would return a different builder so you're not able to call `expectedExceptionMessage` and `expectedTypeOfException` on it.

For the case of multiple constructors we could do `expectMultipleWithParameters(`.

--- a/checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java
+++ b/checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java
@@ -3,6 +3,7 @@ package com.pushtorefresh.private_constructor_checker;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Modifier;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 
@@ -16,25 +17,37 @@ public class PrivateConstructorChecker {
 
     private final String expectedExceptionMessage;
 
+    private Class[] expectedParameters;
+
     private PrivateConstructorChecker(List<Class> classes,
                                       Class<? extends Throwable> expectedTypeOfException,
-                                      String expectedExceptionMessage) {
+                                      String expectedExceptionMessage,
+                                      Class[] expectedParameters) {
         this.classes = classes;
         this.expectedTypeOfException = expectedTypeOfException;
         this.expectedExceptionMessage = expectedExceptionMessage;
+        this.expectedParameters = expectedParameters;
     }
 
-    public static class Builder {
+    public interface Checkable {
+        /**
+         * Runs the check which will assert that required class has one private constructor
+         * which throws or not throws exception and may or may not have expected parameters.
+         */
+        void check();
+    }
 
-        private final List<Class> classes;
-
-        private Class<? extends Throwable> expectedTypeOfException;
-
-        private String expectedExceptionMessage;
-
-        Builder(List<Class> classes) {
-            this.classes = classes;
-        }
+    public interface ExceptionCheckable extends Checkable {
+        /**
+         * Sets the expected message of the exception that must be thrown by the constructor of required class.
+         * <p>
+         * If you don't want to check the type of the exception, you can set just a message.
+         *
+         * @param expectedExceptionMessage message of the exception that must be thrown by the constructor
+         *                                 of required class, should not be {@code null}.
+         * @return ExceptionCheckAble.
+         */
+        ExceptionCheckable expectedExceptionMessage(String expectedExceptionMessage);
 
         /**
          * Sets the expected type of exception that must be thrown by the constructor of required class.
@@ -43,9 +56,38 @@ public class PrivateConstructorChecker {
          *
          * @param expectedTypeOfException type of the exception that must be thrown by the constructor
          *                                of required class, should not be {@code null}.
-         * @return builder.
+         * @return ExceptionCheckAble.
          */
-        public Builder expectedTypeOfException(Class<? extends Throwable> expectedTypeOfException) {
+        ExceptionCheckable expectedTypeOfException(Class<? extends Throwable> expectedTypeOfException);
+    }
+
+    public interface ParametersCheckable extends Checkable {
+        /**
+         * Sets the expected parameters that must match the constructor parameters of required class.
+         *
+         * @param parameters type of parameters that must match the constructor parameters
+         *                   of required class, should not be {@code null}.
+         * @return ExceptionCheckAble.
+         */
+        ParametersCheckable expectedWithParameters(final Class<?>... parameters);
+    }
+
+    public static class Builder implements ExceptionCheckable, ParametersCheckable {
+
+        private final List<Class> classes;
+
+        private Class<? extends Throwable> expectedTypeOfException;
+
+        private String expectedExceptionMessage;
+
+        private Class[] expectedParameters;
+
+        Builder(List<Class> classes) {
+            this.classes = classes;
+        }
+
+        @Override
+        public ExceptionCheckable expectedTypeOfException(Class<? extends Throwable> expectedTypeOfException) {
             if (expectedTypeOfException == null) {
                 throw new IllegalArgumentException("expectedTypeOfException can not be null");
             }
@@ -54,16 +96,8 @@ public class PrivateConstructorChecker {
             return this;
         }
 
-        /**
-         * Sets the expected message of the exception that must be thrown by the constructor of required class.
-         * <p>
-         * If you don't want to check the type of the exception, you can set just a message.
-         *
-         * @param expectedExceptionMessage message of the exception that must be thrown by the constructor
-         *                                 of required class, should not be {@code null}.
-         * @return builder.
-         */
-        public Builder expectedExceptionMessage(String expectedExceptionMessage) {
+        @Override
+        public ExceptionCheckable expectedExceptionMessage(String expectedExceptionMessage) {
             if (expectedExceptionMessage == null) {
                 throw new IllegalArgumentException("expectedExceptionMessage can not be null");
             }
@@ -72,25 +106,33 @@ public class PrivateConstructorChecker {
             return this;
         }
 
-        /**
-         * Runs the check which will assert that required class has one private constructor
-         * which throws or not throws exception.
-         */
+        @Override
+        public ParametersCheckable expectedWithParameters(final Class<?>... expectedParameters) {
+            if (expectedParameters == null) {
+                throw new IllegalArgumentException("expectedParameters can not be null");
+            }
+
+            this.expectedParameters = expectedParameters;
+            return this;
+        }
+
+        @Override
         public void check() {
             new PrivateConstructorChecker(
                     classes,
                     expectedTypeOfException,
-                    expectedExceptionMessage
+                    expectedExceptionMessage,
+                    expectedParameters
             ).check();
         }
     }
 
     /**
-     * Creates instance of {@link com.pushtorefresh.private_constructor_checker.PrivateConstructorChecker.Builder}.
+     * Creates instance of {@link Builder}.
      *
      * @param clazz class that needs to be checked.
-     * @return {@link com.pushtorefresh.private_constructor_checker.PrivateConstructorChecker.Builder} which will prepare
-     * check of the passed class.
+     * @return {@link Builder} which will prepare
+     *         check of the passed class.
      */
     public static Builder forClass(Class clazz) {
         if (clazz == null) {
@@ -138,14 +180,24 @@ public class PrivateConstructorChecker {
 
         final Constructor<?> constructor = constructors[0];
 
-        if (constructor.getParameterTypes().length > 0) {
-            throw new AssertionError(clazz + " has non-default constructor with some parameters");
-        }
-
         constructor.setAccessible(true);
 
         if (!Modifier.isPrivate(constructor.getModifiers())) {
             throw new AssertionError("Constructor of " + clazz + " must be private");
+        }
+
+        final Class<?>[] parameterTypes = constructor.getParameterTypes();
+
+        if (parameterTypes.length > 0) {
+            if (expectedParameters == null) {
+                throw new AssertionError(clazz + " has non-default constructor with some parameters");
+            } else {
+                if (!Arrays.equals(parameterTypes, expectedParameters)) {
+                    throw new AssertionError("Expected constructor with parameters " + getReadableClassesOutput(expectedParameters) + " but found constructor with parameters " + getReadableClassesOutput(parameterTypes));
+                } else {
+                    return;
+                }
+            }
         }
 
         try {
@@ -161,18 +213,11 @@ public class PrivateConstructorChecker {
             // It's okay case if we expect some exception from this constructor
             if (expectedTypeOfException != null || expectedExceptionMessage != null) {
                 if (expectedTypeOfException != null && !expectedTypeOfException.equals(cause.getClass())) {
-                    throw new IllegalStateException("For " + clazz + " expected exception of type = "
-                            + expectedTypeOfException + ", but was exception of type = "
-                            + e.getCause().getClass()
-                    );
+                    throw new IllegalStateException("For " + clazz + " expected exception of type = " + expectedTypeOfException + ", but was exception of type = " + e.getCause().getClass());
                 }
 
                 if (expectedExceptionMessage != null && !expectedExceptionMessage.equals(cause.getMessage())) {
-                    throw new IllegalStateException("For " + clazz + " expected exception message = '"
-                            + expectedExceptionMessage + "', but was = '"
-                            + cause.getMessage() + "'",
-                            e.getCause()
-                    );
+                    throw new IllegalStateException("For " + clazz + " expected exception message = '" + expectedExceptionMessage + "', but was = '" + cause.getMessage() + "'", e.getCause());
                 }
 
                 // Everything is okay
@@ -182,5 +227,21 @@ public class PrivateConstructorChecker {
         } catch (IllegalArgumentException e) {
             throw new RuntimeException("Looks like constructor of " + clazz + " is not default", e);
         }
+    }
+
+    private static String getReadableClassesOutput(final Class<?>[] classes) {
+        final StringBuilder stringBuilder = new StringBuilder("(");
+
+        if (classes.length > 0) {
+            final String concatenator = ", ";
+
+            for (final Class<?> clazz : classes) {
+                stringBuilder.append(clazz.getSimpleName()).append(concatenator);
+            }
+
+            stringBuilder.setLength(stringBuilder.length() - concatenator.length());
+        }
+
+        return stringBuilder.append(')').toString();
     }
 }

--- a/checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java
+++ b/checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java
@@ -156,7 +156,7 @@ public class PrivateConstructionCheckerTest {
 
     @Test
     public void shouldThrowExceptionBecauseTypeOfExpectedExceptionDoesNotMatch() {
-        PrivateConstructorChecker.Builder builder = PrivateConstructorChecker
+        PrivateConstructorChecker.ExceptionCheckable builder = PrivateConstructorChecker
                 .forClass(ClassWithConstructorThatThrowsException.class)
                 .expectedTypeOfException(IllegalArgumentException.class); // Incorrect type
 
@@ -168,7 +168,7 @@ public class PrivateConstructionCheckerTest {
 
     @Test
     public void shouldThrowExceptionBecauseExpectedMessageDoesNotMatch() {
-        PrivateConstructorChecker.Builder builder =PrivateConstructorChecker
+        PrivateConstructorChecker.ExceptionCheckable builder =PrivateConstructorChecker
                 .forClass(ClassWithConstructorThatThrowsException.class)
                 .expectedTypeOfException(IllegalStateException.class) // Correct type
                 .expectedExceptionMessage("lol, not something that you've expected?"); // Incorrect message
@@ -248,5 +248,141 @@ public class PrivateConstructionCheckerTest {
         expectedException.expect(AssertionError.class);
         expectedException.expectMessage(ClassWithoutDefaultConstructor.class + " has non-default constructor with some parameters");
         builder.check();
+    }
+
+    static class ClassWithPrivateStringConstructor {
+        private ClassWithPrivateStringConstructor(String string) {
+
+        }
+    }
+
+    @Test
+    public void shouldCheckPrivateConstructorMatchesParameters() {
+        PrivateConstructorChecker
+                .forClass(ClassWithPrivateStringConstructor.class)
+                .expectedWithParameters(String.class)
+                .check();
+    }
+
+    @Test
+    public void shouldCheckDefaultPrivateConstructorMatchesParameters() {
+        PrivateConstructorChecker
+                .forClass(ClassWithPrivateConstructor.class)
+                .expectedWithParameters()
+                .check();
+    }
+
+    @Test
+    public void shouldThrowExceptionBecauseConstructorParametersDoNotMatch() {
+        final PrivateConstructorChecker.ParametersCheckable builder = PrivateConstructorChecker
+                .forClass(ClassWithPrivateStringConstructor.class)
+                .expectedWithParameters(int.class);
+
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage("Expected constructor with parameters (int) but found constructor with parameters (String)");
+        builder.check();
+    }
+
+    static class ClassWithPrivateStringStringIntConstructor {
+        private ClassWithPrivateStringStringIntConstructor(String string, String str, int value) {
+
+        }
+    }
+
+    @Test
+    public void shouldThrowExceptionBecauseOneConstructorParametersDoesNotMatch() {
+        final PrivateConstructorChecker.ParametersCheckable builder = PrivateConstructorChecker
+                .forClass(ClassWithPrivateStringStringIntConstructor.class)
+                .expectedWithParameters(String.class, String.class, String.class);
+
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage("Expected constructor with parameters (String, String, String) but found constructor with parameters (String, String, int)");
+        builder.check();
+    }
+
+    static class AnotherClassWithPrivateStringConstructor {
+        private AnotherClassWithPrivateStringConstructor(String string) {
+
+        }
+    }
+
+    @Test
+    public void shouldCheckPrivateConstructorMatchesParametersForClasses() {
+        PrivateConstructorChecker
+                .forClasses(ClassWithPrivateStringConstructor.class, AnotherClassWithPrivateStringConstructor.class)
+                .expectedWithParameters(String.class)
+                .check();
+    }
+
+    static class ClassWithProtectedStringConstructor {
+        protected ClassWithProtectedStringConstructor(String string) {
+
+        }
+    }
+
+    @Test
+    public void shouldThrowExceptionBecauseConstructorHasProtectedModifierWhenExceptingParameters() {
+        final PrivateConstructorChecker.ParametersCheckable builder = PrivateConstructorChecker
+                .forClass(ClassWithProtectedStringConstructor.class)
+                .expectedWithParameters(String.class);
+
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage("Constructor of " + ClassWithProtectedStringConstructor.class + " must be private");
+        builder.check();
+    }
+
+    static class ClassWithPublicStringConstructor {
+        public ClassWithPublicStringConstructor(String string) {
+
+        }
+    }
+
+    @Test
+    public void shouldThrowExceptionBecauseConstructorHasPublicModifierWhenExceptingParameters() {
+        final PrivateConstructorChecker.ParametersCheckable builder = PrivateConstructorChecker
+                .forClass(ClassWithPublicStringConstructor.class)
+                .expectedWithParameters(String.class);
+
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage("Constructor of " + ClassWithPublicStringConstructor.class + " must be private");
+        builder.check();
+    }
+
+    static class ClassWithDefaultStringConstructor {
+        ClassWithDefaultStringConstructor(String string) {
+
+        }
+    }
+
+    @Test
+    public void shouldThrowExceptionBecauseConstructorHasDefaultModifierWhenExceptingParameters() {
+        final PrivateConstructorChecker.ParametersCheckable builder = PrivateConstructorChecker
+                .forClass(ClassWithDefaultStringConstructor.class)
+                .expectedWithParameters(String.class);
+
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage("Constructor of " + ClassWithDefaultStringConstructor.class + " must be private");
+        builder.check();
+    }
+
+    @Test
+    public void shouldThrowExceptionIfMoreThanOneConstructorExistsWhenUsingExpectWithParameters() {
+        final PrivateConstructorChecker.ParametersCheckable builder = PrivateConstructorChecker
+                .forClass(ClassWith2Constructors.class)
+                .expectedWithParameters(String.class);
+
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage(ClassWith2Constructors.class + " has more than one constructor");
+        builder.check();
+    }
+
+    @Test
+    public void builderShouldThrowExceptionIfNullWasPassedAsExpectedParameters() {
+        final PrivateConstructorChecker.ParametersCheckable builder = PrivateConstructorChecker
+                .forClass(ClassWithPrivateConstructor.class);
+
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("expectedParameters can not be null");
+        builder.expectedWithParameters(null);
     }
 }

